test/fuzz/utils.hpp

```diff
@@ -0,0 +1,219 @@
+// Copyright (C) 2023 Intel Corporation
+// Part of the Unified-Runtime Project, under the Apache License v2.0 with LLVM Exceptions.
+// See LICENSE.TXT
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include <fstream>
+#include <iostream>
+#include <iterator>
+#include <map>
+#include <vector>
+
+namespace fuzz {
+
+enum FuzzerAPICall {
+    UR_PLATFORM_GET,
+    UR_DEVICE_GET,
+    UR_DEVICE_RELEASE,
+    UR_CONTEXT_CREATE,
+    UR_CONTEXT_RELEASE,
+    UR_USM_POOL_CREATE_HOST,
+    UR_USM_POOL_CREATE_DEVICE,
+    UR_USM_POOL_RELEASE_HOST,
+    UR_USM_POOL_RELEASE_DEVICE,
+    UR_USM_HOST_ALLOC_POOL,
+    UR_USM_HOST_ALLOC_NO_POOL,
+    UR_USM_DEVICE_ALLOC_POOL,
+    UR_USM_DEVICE_ALLOC_NO_POOL,
+    UR_USM_FREE_HOST_POOL,
+    UR_USM_FREE_HOST_NO_POOL,
+    UR_USM_FREE_DEVICE_POOL,
+    UR_USM_FREE_DEVICE_NO_POOL,
+    UR_PROGRAM_CREATE_WITH_IL,
+    UR_MAX_FUZZER_API_CALL,
+};
+
+typedef struct FuzzerInput {
+    uint8_t *data_ptr;
+    size_t data_size;
+} FuzzerInput;
+
+typedef struct TestState {
+    static constexpr uint32_t num_entries = 1;
+
+    FuzzerInput input;
+    ur_loader_config_handle_t config;
+
+    std::vector<ur_adapter_handle_t> adapters;
+    std::vector<ur_platform_handle_t> platforms;
+    std::vector<ur_device_handle_t> devices;
+    std::vector<ur_context_handle_t> contexts;
+    std::map<ur_usm_pool_handle_t, std::vector<void *>> pool_host_allocs;
+    std::map<ur_usm_pool_handle_t, std::vector<void *>> pool_device_allocs;
+    std::vector<void *> no_pool_host_allocs;
+    std::vector<void *> no_pool_device_allocs;
+    ur_device_type_t device_type = UR_DEVICE_TYPE_ALL;
+
+    uint32_t num_adapters;
+    uint32_t num_platforms;
+    uint32_t num_devices;
+
+    uint8_t platform_num;
+    uint8_t device_num;
+    uint8_t context_num;
+    uint8_t device_type_fuzz;
+} TestState;
+
+//////////////////////////////////////////////////////////////////////////////
+template <typename T> int get_next_input_data(FuzzerInput *input, T *out_data) {
+    size_t out_data_size = sizeof(out_data);
+    if (input->data_size == 0 || input->data_size < out_data_size) {
+        return -1;
+    }
+    *out_data = *input->data_ptr;
+    input->data_ptr += out_data_size;
+    input->data_size -= out_data_size;
+
+    return 0;
+}
+
+int init_random_data(TestState *state) {
+    if (state->input.data_size < 5) {
```

Comment:
hm, wouldn't it be better to check result of `get_next_input_data`? otherwise, if the size of these fields changes, this will not be correct.

Reply:
I replaced multiple checks after each and every `get_next_input_data` with this look-ahead check for all values at once. I will change it though, as one have to remember to change this value every time a "random" var is added/removed.